### ISSUE TITLE
Recover git history from fluxcd/gitops-working-group

### ISSUE THREAD

**Issue by scottrigby** [gitops-working-group/gitops-working-group#37]
## Problem

The initial goal of fluxcd/gitops-working-group was a temporary home for the WG, and after a time it would be moved to a new org. The repo was copied though rather than moved, and git history was lost (including contributions from members of different orgs, which we would like to preserve).

- See https://github.com/gitops-working-group/gitops-working-group-old/issues/49

## Proposed Solution

Copy git history from fluxcd/gitops-working-group, and replay new change commits from this new repo on top of that existing history. The result should be the exact same content we have now, but with history preserved from the old repo.
- The deprecation notice in the old repo should not be brought over https://github.com/gitops-working-group/gitops-working-group-old/commit/ccce85746ece9cb825691b874bebc634e1bf244b
- The initial commit that copied content from the old repo into this new repo (#1) should be converted to a diff on top of the code what was already there
- The initial commit in the new repo should be omitted, to preserve continuity of history https://github.com/gitops-working-group/gitops-working-group/commit/1733ae33f79b714b1ed42ca240a6b1f6d3161533

## Progress

I'll list the exact steps, so everything is transparent for any users who wish to follow along at home 😄 :

- [x] Clone new and add old remote, named for clarity

    ```console
    $ git clone --origin new git@github.com:gitops-working-group/gitops-working-group.git
    $ cd gitops-working-group/
    $ git remote add old git@github.com:gitops-working-group/gitops-working-group-old.git
    $ git remote -v
    new	git@github.com:gitops-working-group/gitops-working-group.git (fetch)
    new	git@github.com:gitops-working-group/gitops-working-group.git (push)
    old	git@github.com:gitops-working-group/gitops-working-group-old.git (fetch)
    old	git@github.com:gitops-working-group/gitops-working-group-old.git (push)
    ```

- [x] Set up a branch to combine the history

    ```console
    $ git fetch --all
    $ git checkout -b combine-history
    Switched to a new branch 'combine-history'
    ```

- [x] Start with the old repo history since that should be first

    ```console
    $ git reset --hard old/main 
    HEAD is now at ccce857 Add deprecation notice
    ```

- [x] Remove old repo deprecation notice by resetting to the current SHA's parent commit ('%P')

    ```console
    $ git reset --hard $(git log -1 --pretty='%P')
    HEAD is now at 66871ac Merge pull request #44 from scottrigby/slack-and-markdownlint
    ```

- [x] Get all commits from new repo, except the initial commit, in order to preserve continuity of history. For the end result we want the initial copy commit (from #1) to be converted into a diff on top of existing code from commits in the old repo.

    ```console
    $ new_commits=($(git rev-list new/main | tac | tail -n +2))
    $ for i in "${new_commits[@]}"; do git cherry-pick $i --strategy-option theirs; done
    Auto-merging README.md
    CONFLICT (add/add): Merge conflict in GOVERNANCE.md
    Auto-merging GOVERNANCE.md
    CONFLICT (add/add): Merge conflict in CODE_OF_CONDUCT.md
    Auto-merging CODE_OF_CONDUCT.md
    [combine-history 035d688] Move content from fluxcd/gitops-working-group (#1)
     Author: Cornelia Davis <cornelia@corneliadavis.com>
     Date: Tue Dec 15 10:42:28 2020 -0800
     3 files changed, 161 insertions(+), 3 deletions(-)
     rewrite GOVERNANCE.md (100%)
    [combine-history 27c6858] Update MEETINGS.md (#3)
     Author: Daniel Lizio-Katzen <daniel@djlk.com>
     Date: Thu Dec 17 22:18:51 2020 -0500
     1 file changed, 8 insertions(+), 2 deletions(-)
    [combine-history c7722a5] Updated with Feb meeting & Jan recording/notes (#35)
     Author: Daniel Lizio-Katzen <daniel@djlk.com>
     Date: Mon Jan 18 10:53:52 2021 -0500
     1 file changed, 10 insertions(+), 4 deletions(-)
    ```

- [x] We should now be good to go. We can see Cornelia's first commit in this new repo is now a proper diff from what was present in the old repo:

    ```console
    $ git log --oneline
    c7722a5 (HEAD -> combine-history) Updated with Feb meeting & Jan recording/notes (#35)
    27c6858 Update MEETINGS.md (#3)
    035d688 Move content from fluxcd/gitops-working-group (#1)
    66871ac Merge pull request #44 from scottrigby/slack-and-markdownlint
    ...
    $ git show 035d688
    ```
    <details>
    <summary>CLICK TO EXPAND DIFF</summary>

    ```diff
    git show 035d688
    commit 035d68809191b10d6c119b3f4d7323e23d28ed85
    Author: Cornelia Davis <cornelia@corneliadavis.com>
    Date:   Tue Dec 15 10:42:28 2020 -0800
    
        Move content from fluxcd/gitops-working-group (#1)
        
        * Move content over from fluxcd/gitops-working-group
        
        * Update git URLs to reflect new home
    
    diff --git a/CODE_OF_CONDUCT.md b/CODE_OF_CONDUCT.md
    index 153bab7..69c7d6b 100644
    --- a/CODE_OF_CONDUCT.md
    +++ b/CODE_OF_CONDUCT.md
    @@ -1,3 +1,3 @@
     # Code of Conduct
     
    -The Flux community follows the [CNCF Code of Conduct](https://github.com/cncf/foundation/blob/master/code-of-conduct.md).
    +The GitOps Working Group community follows the [CNCF Code of Conduct](https://github.com/cncf/foundation/blob/master/code-of-conduct.md).
    diff --git a/GOVERNANCE.md b/GOVERNANCE.md
    index 73c06c5..4370d1e 100644
    --- a/GOVERNANCE.md
    +++ b/GOVERNANCE.md
    @@ -1 +1,159 @@
    -The GitOps Working Group follows the governance process described in https://github.com/fluxcd/community/blob/main/GOVERNANCE.md.
    +<!-- see https://github.com/yzhang-gh/vscode-markdown/blob/master/README.md#table-of-contents -->
    +<!-- omit in toc -->
    +# GitOps Working Group Governance
    +
    +This document <https://github.com/gitops-working-group/gitops-working-group/blob/main/GOVERNANCE.md> defines the governance process for the GitOps Working Group project and community.
    +
    +- [Values](#values)
    +  - [Code of Conduct](#code-of-conduct)
    +- [Roles in the GitOps Working Group Community](#roles-in-the-gitops-working-group-community)
    +  - [Users](#users)
    +  - [Contributors](#contributors)
    +  - [Maintainers](#maintainers)
    +  - [Oversight Committee](#oversight-committee)
    +- [Decision Making](#decision-making)
    +  - [Deciders](#deciders)
    +  - [Decision Guidelines](#decision-guidelines)
    +  - [Simple Majority Decisions](#simple-majority-decisions)
    +  - [Supermajority Decisions](#supermajority-decisions)
    +  - [Unanimity Decisions](#unanimity-decisions)
    +- [Proposal Process](#proposal-process)
    +- [Licenses and Copyright](#licenses-and-copyright)
    +
    +## Values
    +
    +- **Open:**
    +The GitOps Working Group community strives to be open, accessible and welcoming to everyone.
    +Anyone may contribute, and contributions are available to all users according to open source values and licenses.
    +- **Transparent:**
    +The GitOps Working Group strives for transparency in all discussions, announcements, disclosures and decision making.
    +- **Unbiased:**
    +The GitOps Working Group strives to operate independently of specific partisan interests, and for decision making to fairly balance the wider community interests of its end users and contributors.
    +
    +### Code of Conduct
    +
    +The GitOps Working Group community adheres to the CNCF Code of Conduct <https://github.com/cncf/foundation/blob/master/code-of-conduct.md>.
    +
    +Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting a _GitOps Working Group_ Oversight Committee member.
    +
    +If no conclusion can be reached in meditation, such issues can be escalated to the CNCF mediator, Mishi Choudhary <mishi@linux.com>, in which case CNCF may choose to intervene.
    +
    +## Roles in the GitOps Working Group Community
    +
    +The GitOps Working Group community comprises the following roles:
    +
    +### Users
    +
    +GitOps end users are the most important aspect of the community, without whom the project would have no purpose. Users are anyone who has a need for the project.
    +Apart from following the Code of Conduct, there are no special requirements.
    +
    +### Contributors
    +
    +The GitOps Working Group welcomes all kinds of contributions, including code, issues, documentation, external tools, advocacy and community work.
    +As a contributor we want to invite you to join the discussions in a variety of forums laid out in TBD.
    +
    +### Maintainers
    +
    +Maintainers are elected Contributors who showed significant and sustained contributions in a git repository.
    +The initial set of maintainers are drawn from the organizations that proposed the creation of the working group.
    +Current Maintainers are listed in a `MAINTAINERS` file at the root of this git repository.
    +
    +Maintainers are expected to:
    +
    +- Enable and promote GitOps Working Group community values
    +- Engage with end Users through appropriate communication channels
    +- Serve as a point of conflict resolution between Contributors to their git repository
    +- Maintain open collaboration with Contributors and other Maintainers
    +- Ask for help when unsure and step down considerately
    +
    +Maintainers will be invited to the `@gitops-working-group/maintainers` <https://github.com/orgs/gitops-working-group/teams/maintainers> team, and are members of this team for as long as they are involved with the project.
    +
    +### Oversight Committee
    +
    +This committee is responsible for the overall project, and anything not easily managed by the Maintainers of each git repository. Including:
    +
    +- Overseeing the project health and growth
    +- Maintaining the brand, mission, vision, values, and scope of the overall project
    +- Changes to licensing and intellectual property
    +- Administering access to all project assets
    +- Administering git repositories as needed
    +- Handling Code of Conduct violations
    +- Managing financial decisions
    +- Defining the scope of each git repository
    +- Resolving escalated decisions when Maintainers responsible are blocked
    +
    +Ultimately the committee - after consulting with the collective of Maintainers and their community - drive the direction, values and governance of the overall project.
    +
    +This committee will initially be comprised of GitOps Working Group Maintainers who have steered the project prior to this initial Governance document.
    +The aspiration is no one company or organization should have oversight of the overall project, however that is not yet realistic at this stage. The goal is to broaden maintainership to include a wider range of organizations during CNCF incubation.
    +
    +Oversight Committee members are publicly listed in the `@gitops-working-group/oversight-committee` <https://github.com/orgs/gitops-working-group/teams/oversight-committee> GitHub team.
    +
    +## Decision Making
    +
    +### Deciders
    +
    +- Repository Maintainers: Decisions that affect only one git repository.
    +- Oversight Committee: Decisions that are outside the scope of a single git repository.
    +
    +### Decision Guidelines
    +
    +- Decisions that warrant wider input should be made public by using the below guidelines in combination with the Proposal Process below.
    +- Whether or not wider input is required, the GitOps Working Group community believes that the best decisions are reached through Consensus <https://en.wikipedia.org/wiki/Consensus_decision-making>.
    +- Most decisions start by seeking Lazy Consensus <https://communitymgt.wikia.com/wiki/Lazy_consensus>.
    +- If an objection is raised through the Lazy Consensus process, Deciders work together to seek an agreeable solution.
    +- If Consensus can not be reached, but a decision must be made, the next step is try to attempt to agree that a vote should be called.
    +  This is important, as it gives dissenting views a chance to request more information or raise further points.
    +  If Deciders are the Oversight Committee, part of that responsibility is the final point of escalation, so agreeing to a vote is assumed if timeline doesn't allow the consensus process to continue.
    +- If Deciders are Repository Maintainers, and they can't agree on calling a vote, they may escalate to the Oversight Committee.
    +  This should only be done at this stage if:
    +  1. An unmovable deadline is threatened by continuing the Consensus process; or
    +  2. A Decider feels there is unreasonable blocking of both reaching Consensus and agreeing to a vote.
    +      This should be rare, due to the social cost of discontinuing the Consensus process for this reason.
    +      Most decisions should wait for the above process to take its course.
    +- If Deciders agree to a vote, the default is a Simple Majority.
    +- However, there are cases that require stronger voting – Supermajority or Unanimity – specified below:
    +
    +### Simple Majority Decisions
    +
    +If a vote is called, the default is a Simple Majority Vote <https://en.wikipedia.org/wiki/Majority>.
    +
    +### Supermajority Decisions
    +
    +If a vote is called, the following decisions require a Supermajority Vote <https://en.wikipedia.org/wiki/Supermajority>.
    +
    +- Oversight Committee: Enforcing a Code of Conduct violation by a community member.
    +- Oversight Committee: Licensing and intellectual property changes.
    +- Oversight Committee: Material changes to the Governance document.
    +  - Note: editorial changes to governance may be made by lazy consensus, unless challenged.
    +    These are changes which fix spelling or grammar, update work affiliation or similar, update style or reflect an outside and obvious reality.
    +    They do not change the intention or meaning of anything in this document.
    +
    +### Unanimity Decisions
    +
    +If a vote is called, the following decisison require Unanimity <https://en.wikipedia.org/wiki/Unanimity>.
    +
    +- Repository Maintainers: Electing new Maintainers of the same repository.
    +- Oversight Committee: Electing new Committee members.
    +- Oversight Committee: Removing a Repository Maintainer or Committee member for any reason other than inactivity.
    +
    +## Proposal Process
    +
    +- Changes to artifacts produced by the GitOps Working Group (documentation, samples, etc.) should go through the pull request process, where the idea and any implementation details can be publicly discussed with Maintainers, other contributors, and end users.
    +  Pull requests should only be merged after receiving GitHub approval from at least one Maintainer who is not the pull request author.
    +- Non-artifact changes should be proposed as GitHub issues.
    +  If unclear which git repository to create the issue in, default to the community repository <https://github.com/gitops-working-group/community>.
    +- All proposals should be discussed publicly in an appropriate GitHub issue or pull request.
    +- If a Maintainer of an affected git repository feels feedback from specific people is warranted they will @mention those users or teams to request feedback.
    +- Proposals may also be added to the GitOps Working Group meetings agenda, as a good avenue for making progress on a decision <https://lists.cncf.io/g/cncf-gitops-working-group/calendar>.
    +
    +## Licenses and Copyright
    +
    +- Apache 2.0 is required for all git repositories.
    +- Developer Certificate of Origin (DCO) commit signoff is required for all new code contributions.
    +
    +Links to relevant CNCF documentation:
    +
    +- <https://github.com/cncf/foundation/blob/master/charter.md#11-ip-policy>
    +- <https://github.com/cncf/foundation/blob/master/allowed-third-party-license-policy.md#approved-licenses-for-allowlist>
    +- <https://github.com/cncf/foundation/blob/master/copyright-notices.md#copyright-notices>
    diff --git a/README.md b/README.md
    index ad63ace..21e95ae 100644
    --- a/README.md
    +++ b/README.md
    @@ -2,7 +2,7 @@
     
     ## Announcing the GitOps Working Group
     
    -On November 19, 2020, Amazon, Codefresh, GitHub, Microsoft, and Weaveworks are announcing the creation of the GitOps Working Group. This will be an open CNCF community project created inside the CNCF [fluxcd GitHub organization](https://github.com/fluxcd/flux) as the initial venue for collaboration and open governance.
    +On November 19, 2020, Amazon, Codefresh, GitHub, Microsoft, and Weaveworks are announcing the creation of the GitOps Working Group. This will be an open CNCF community project created inside the CNCF [gitops-working-group GitHub organization](https://github.com/gitops-working-group) as the initial venue for collaboration and open governance.
     
     The GitOps Working Group’s goal is to provide companies and individuals with the skills, knowledge and competency to implement GitOps tooling and methodologies which simplify the operation and management of infrastructure and cloud native applications.
     ```
    </details>

- [x] Push new branch to this repo for review/approval

    ```console
    $ git push new combine-history
    ```

- [x] For posterity, also add a backup of the current `main` to keep around for a time in case there are any questions.
    
    ```console
    $ git checkout main
    $ git checkout -b main-backup
    $ git push new main-backup
    ```

- [x] Review/approval. Because the next step is a destructive move, I would like one manual review of my above steps and the `combine-history` branch before we replace `main` with it.

    @bricef please review. You can see the result here, which should be exactly what we ultimately want for `main`:
    https://github.com/gitops-working-group/gitops-working-group/commits/combine-history

- [x] Force-push `combine-history` to overwrite `main` in this repo. We can't accomplish this through a PR because the old and new repos will have "unrelated history" until after the force-push is complete.

- [x]  Once the previous step is complete, any existing pull requests will need to be re-created. Git users could cherry pick their existing PR commit(s) on top of the new `main`, and open a new PR. However we are just going ahead and doing this for them (@bricef and @scottrigby)

- [x] Announce change in CNCF `#wg-gitops` Slack channel (done, see https://cloud-native.slack.com/archives/C01G9DEE88M/p1611229997007300)

- [x] Add change announcement to agenda for next WG meeting (done, added to https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk)

**Comment by scottrigby** [comment]
## Steps

- [ ] clone new and add old remote, named for clarity

    ```console
    $ git clone --origin new git@github.com:gitops-working-group/gitops-working-group.git
    $ cd gitops-working-group/
    $ git remote add old git@github.com:gitops-working-group/gitops-working-group-old.git
    $ git remote -v
    new	git@github.com:gitops-working-group/gitops-working-group.git (fetch)
    new	git@github.com:gitops-working-group/gitops-working-group.git (push)
    old	git@github.com:gitops-working-group/gitops-working-group-old.git (fetch)
    old	git@github.com:gitops-working-group/gitops-working-group-old.git (push)
    ```

- [ ] Set up a branch to combine the history

    ```console
    $ git fetch --all
    $ git checkout -b combine-history
    Switched to a new branch 'combine-history'
    ```

- [ ] Start with the old repo history since that should be first

    ```console
    $ git reset --hard old/main 
    HEAD is now at ccce857 Add deprecation notice
    ```

- [ ] Remove old repo deprecation notice by resetting to the current SHA's parent commit ('%P')

    ```console
    $ git reset --hard $(git log -1 --pretty='%P')
    HEAD is now at 66871ac Merge pull request #44 from scottrigby/slack-and-markdownlint
    ```

- [ ] Get all commits from new repo, except the initial commit, in order to preserve continuity of history. For the end result we want the initial commit that copied content to this new repo (from #1) to be converted to a diff on top of code that was already present from commits in the old repo.

    ```console
    $ new_commits=($(git rev-list new/main | tac | tail -n +2))
    $ for i in "${new_commits[@]}"; do git cherry-pick $i --strategy-option theirs; done
    Auto-merging README.md
    CONFLICT (add/add): Merge conflict in GOVERNANCE.md
    Auto-merging GOVERNANCE.md
    CONFLICT (add/add): Merge conflict in CODE_OF_CONDUCT.md
    Auto-merging CODE_OF_CONDUCT.md
    [combine-history 035d688] Move content from fluxcd/gitops-working-group (#1)
     Author: Cornelia Davis <cornelia@corneliadavis.com>
     Date: Tue Dec 15 10:42:28 2020 -0800
     3 files changed, 161 insertions(+), 3 deletions(-)
     rewrite GOVERNANCE.md (100%)
    [combine-history 27c6858] Update MEETINGS.md (#3)
     Author: Daniel Lizio-Katzen <daniel@djlk.com>
     Date: Thu Dec 17 22:18:51 2020 -0500
     1 file changed, 8 insertions(+), 2 deletions(-)
    [combine-history c7722a5] Updated with Feb meeting & Jan recording/notes (#35)
     Author: Daniel Lizio-Katzen <daniel@djlk.com>
     Date: Mon Jan 18 10:53:52 2021 -0500
     1 file changed, 10 insertions(+), 4 deletions(-)
    ```

- [ ] We should now be good to go. We can see Cornelia's first commit in the new repo is now a proper diff from what was present in the old repo:

    ```console
    $ git log --oneline
    c7722a5 (HEAD -> combine-history) Updated with Feb meeting & Jan recording/notes (#35)
    27c6858 Update MEETINGS.md (#3)
    035d688 Move content from fluxcd/gitops-working-group (#1)
    66871ac Merge pull request #44 from scottrigby/slack-and-markdownlint
    ...
    $ git show 035d688
    <DIFF>
    ```

- [ ] Push new branch to this repo for review/approval

    ```console
    $ git push new combine-history
    ```
    
    You can see the result here, which is exactly what we want:
    https://github.com/gitops-working-group/gitops-working-group/commits/combine-history

- [ ] Review/approval. Because the next step is a destructive move, I would like one manual review of my above steps and the `combine-history` branch before we replace `main` with it. For posterity, I will also add a backup of the current `main` to keep around for a time in case there are any questions.
    
    ```console
    $ git checkout main
    $ git checkout -b main-backup
    $ git push new main-backup
    ```

- [ ] After review/approval, we must ultimately force-push `combine-history` to overwrite `main` in this repo. We can't accomplish this through a PR because the old and new repos will have "unrelated history" until after the force-push is complete.

    **Note:** Once this step is complete, any existing pull requests will need to be re-created. Git users should be able to cherry pick their existing PR commit(s) on top of the new `main`, and open a new PR.

**Comment by bricef** [comment]
@scottrigby - this sounds good to me. You should be able top carry out the required steps yourself. Please go ahead.

**Comment by bricef** [comment]
I am temporarily removing branch protection to enable the force push.

**Comment by scottrigby** [comment]
It's all good, I've got it 👍 Thanks

**Comment by bricef** [comment]
Done.

**Comment by bricef** [comment]
I am re-enabling main branch protection after force-pushing

**Comment by bricef** [comment]
We'll need to update the outstanding pull requests. Taking a look at this now.

**Comment by bricef** [comment]
I have re-created the two outstanding PRs: https://github.com/gitops-working-group/gitops-working-group/pull/42 and https://github.com/gitops-working-group/gitops-working-group/pull/41

I believe this closes the issue? @scottrigby Anything else outstdanding?

**Comment by scottrigby** [comment]
Ok thanks for the tag-team. I tried to slack between my last commend and this one but maybe your notifications are off? Without real time chat I've held off doing anything in the meantime so we don't cross wires.

**Comment by scottrigby** [comment]
Re PRs, was https://github.com/gitops-working-group/gitops-working-group/pull/34 still in queue to be merged? I'd be happy to re-create that one too that if it helps.

**Comment by bricef** [comment]
@scottrigby Sounds good. Will approve #43 again as well.

**Comment by scottrigby** [comment]
Yep, looks like we're done re-creating the PRs 🎉  

I added 2 more checklist items for announcing the change 👍
